lib/marked.js

```diff
@@ -505,7 +550,13 @@ inline.normal = merge({}, inline);
 
 inline.pedantic = merge({}, inline.normal, {
   strong: /^__(?=\S)([\s\S]*?\S)__(?!_)|^\*\*(?=\S)([\s\S]*?\S)\*\*(?!\*)/,
-  em: /^_(?=\S)([\s\S]*?\S)_(?!_)|^\*(?=\S)([\s\S]*?\S)\*(?!\*)/
+  em: /^_(?=\S)([\s\S]*?\S)_(?!_)|^\*(?=\S)([\s\S]*?\S)\*(?!\*)/,
+  link: edit(/^!?\[(label)\]\(\s*<?([\s\S]*?)>?(?:\s+(['"][\s\S]*?['"]))?\s*\)/)
```

Comment:
This regex is vulnerable.

Reply:
A static scan of the tree on @Feder1co5oave's PR says:

```
"vulnRegexes":["^!?\\[(label)\\]\\(\\s*<?([\\s\\S]*?)>?(?:\\s+(['\"][\\s\\S]*?['\"]))?\\s*\\)"]
```
(JSON-encoded).

Checking the ones requested individually...

Reply:
I'm pretty sure this is the "old" link regex, I didn't change it.
Could you point out where the issue is? I'm thinking this part?

    (?:\s+(['"][\s\S]*?['"]))?

The double ?, maybe? I'm kinda lost.

Reply:
This is what `check-regex.pl` says:

```
{"pattern":"^!?\\[(label)\\]\\(\\s*<?([\\s\\S]*?)>?(?:\\s+(['\"][\\s\\S]*?['\"]))?\\s*\\)",
"evilInput": {"suffix":"e","pumpPairs":[
{"pump":"a","prefix":"![label](\t"},{"prefix":"a","pump":"\t"}]
},
"timeLimit":"5",
"nPumps":"250000",
"language":"javascript"}
```

The issue is the use of `[\s\S]*?` twice.
This results in a quadratic regex.

Reply:
Ok I see where we're getting at... I'm not sure what that output means, can you translate that into some example input that triggers the quadratic behavior?
I was thinking, maybe...?

    ![label](\t\t\t\t\t\t\t\t\t...

Reply:
`![label](\taaaaaaaaaaaaaaaa...aaaaaa\t\t\t\t\t\t\t\t....\t\t\te`

Reply:
Not sure if that string is overly complex or how many pumps you need.

Reply:
I'm not sure how to fix this.

Reply:
Can you try with `/^!?\[(label)\]\((\s*(<[^>]*>))?(?:\s+(['"][\s\S]*?['"]))?\s*\)/` ?

[existing](https://regexper.com/#%2F%5E!%3F%5C%5B%28label%29%5C%5D%5C%28%5Cs*%3C%3F%28%5B%5Cs%5CS%5D*%3F%29%3E%3F%28%3F%3A%5Cs%2B%28%5B'%22%5D%5B%5Cs%5CS%5D*%3F%5B'%22%5D%29%29%3F%5Cs*%5C%29%2F), [proposed](https://regexper.com/#%2F%5E!%3F%5C%5B%28label%29%5C%5D%5C%28%28%5Cs*%28%3C%5B%5E%3E%5D*%3E%29%29%3F%28%3F%3A%5Cs%2B%28%5B'%22%5D%5B%5Cs%5CS%5D*%3F%5B'%22%5D%29%29%3F%5Cs*%5C%29%2F)

Reply:
wrapping the URI in angular parenthesis is optional.
The issue is that in this original rule, spaces where allowed inside the URI, which _could_ then be followed by space and a title wrapped in single/double quotes. There is no easy way of finding where the uri stops and when the title begins, but for the presence of the opening quote (which in some cases could be part of the URI instead?). The rule itself is ambiguous.

Reply:
It doesn't sound like this can be done cleanly in a single regex. Can you provide a link to the spec for this rule?

Reply:
This is the "legacy" link rule of marked, it's been like this for a long time. I didn't change it. It doesn't follow any clear spec, it just tries its best. We should come up with a different rule that doesn't break compatibility.
However, keep in mind this is used only in pedantic mode, I think very few people every used it and it is planned to be deprecated soon.

Reply:
OK. Since you didn't change it, it shouldn't prevent the PR from being merged.